### DISCUSSION TITLE
With Tuist's µFeature Architecture?

### DISCUSSION THREAD

**Discussion by kimxwan0319** [pointfreeco/swift-dependencies#147] [General]
## How can I use swift-dependencies with MFA
Is there anyone who can share my concerns?

<img width="423" alt="Screenshot 2023-11-29 at 3 41 54 PM" src="https://github.com/pointfreeco/swift-dependencies/assets/67373938/20f2b046-d549-4970-bcaf-fef91d381118">

* Due to the characteristics of MFA, interfaces that require external access should be placed in the Interface module, implementations in the Source module, and mock objects needed for testing in the Testing module.
	* Structs defining closures -> Interface module
	* Extensions inheriting `DependencyKey` and implementing `liveValue` -> Source module
	* Extensions inheriting `TestDependencyKey` and implementing `testValue` -> Testing module
* To actually use a Dependency, it must be registered in the DependencyValues extension. **So, who should be responsible for this registration?**
	* **Interface module?** ❌
		* Although it seems best since it only needs to be registered once and is accessible by other services depending on the same service, it's not possible because the Interface module does not implement either `DependencyKey` or `TestDependencyKey`.
		* There's no way to explicitly implement this... live or test Value is a mandatory element.
	* **Source module?**
		* If it wasn't added in the interface, it must be necessarily included in the source module to use the registered dependency.
		* If registered publicly in the source module, tests and example modules depending on both source and testing modules can access all liveData and testData of the registered dependency.
		* ### **But how can other services that do not directly depend on the source of a particular service use its dependencies?**

**Comment by arslaman** [comment]
One way to solve this is to store extensions inheriting TestDependencyKey and implementing testValue in the interface module.

**Comment by shp7724** [comment]
I came across the same question, and I ended up implementing the `liveValue` in the main app target (which can import all feature modules) and extend `DependencyValues` with only `testValue` in the interface module as @arslaman mentioned.

https://swiftpackageindex.com/pointfreeco/swift-dependencies/1.3.0/documentation/dependencies/livepreviewtest#Separating-interface-and-implementation
